kale/evaluate/uncertainty_metrics.py

```diff
@@ -0,0 +1,1036 @@
+    for binidx in range(len(all_qs_perc)):
+        bin_accs = all_qs_perc[binidx]
+        bin_asizes = all_qs_size[binidx]
+
+        weighted_mean_bin = 0.0
+        total_weights_bin = 0.0
+        for l_idx in range(len(bin_accs)):
+            b_acc = bin_accs[l_idx]
+            b_siz = bin_asizes[l_idx]
+            weighted_mean_bin += b_acc * b_siz
+            total_weights_bin += b_siz
+
+        # Avoid div by 0
+        if weighted_mean_bin == 0 or total_weights_bin == 0:
+            weighted_av_bin = 0.0
+        else:
+            weighted_av_bin = weighted_mean_bin / total_weights_bin
+        weighted_av_binwise.append(weighted_av_bin)
+
+    # No weighted average, just normal av
+    normal_av_bin_wise = []
+    for binidx in range(len(all_qs_perc)):
+        bin_accs = all_qs_perc[binidx]
+        normal_av_bin_wise.append(np.mean(bin_accs))
+
+    return {
+        "mean all lms": np.mean(all_lm_perc),
+        "mean all bins": weighted_av_binwise,
+        "mean all": all_qs_perc,
+        "all bins concatenated lms seperated": all_qs_errorbound_concat_lms_sep,
+    }
+
+
+def strip_for_bound(string_: str) -> list:
+    """
+    Convert a string containing comma-separated floats into a list of floats.
+
+    Args:
+        string_ (str): A string containing floats, separated by commas.
+
+    Returns:
+        list: A list of floats.
+
+    Example:
+        >>> strip_for_bound("[1.0, 2.0], [3.0, 4.0]")
+        [[1.0, 2.0], [3.0, 4.0]]
+    """
+    bounds = []
+    for entry in string_:
+        entry = entry[1:-1]
+        bounds.append([float(i) for i in entry.split(",")])
+    return bounds
+
+
+def evaluate_correlations(
+    bin_predictions: Dict[str, pd.DataFrame],
+    uncertainty_error_pairs: List[Tuple[str, str, str]],
+    cmaps: List[Dict[Any, Any]],
+    num_bins: int,
+    conf_invert_info: List[Tuple[str, bool]],
+    num_folds: int = 8,
+    pixel_to_mm_scale: float = 1,
+    combine_middle_bins: bool = False,
+    save_path: Optional[str] = None,
+    to_log: bool = False,
+) -> Dict[str, Dict[str, Dict[str, Any]]]:
+    """
+    Calculates the correlation between error and uncertainty for each bin and for each landmark,
+    using a piece-wise linear regression model.
+
+    Args:
+        bin_predictions: A dictionary of Pandas DataFrames containing model predictions for each testing fold.
+        uncertainty_error_pairs: A list of tuples specifying the names of the uncertainty,
+            error, and uncertainty inversion keys for each pair.
+        cmaps: A dictionary of colour maps to use for plotting the results.
+        num_bins: The number of quantile bins to divide the data into.
+        conf_invert_info: A list of tuples specifying whether to invert the uncertainty values for each method.
+        num_folds: The number of folds to use for cross-validation (default: 8).
+        pixel_to_mm_scale: The scale factor to convert from pixels to millimeters (default: 1).
```

Comment:
Is this generic? Do I have to do this for all use of Qbin?

Reply:
yes it is generic. it works if folds =1 (i.e. not CV). pixel_to_mm is if to scale error, default is no scale. Conf_invert is whether it is confidence or uncertainty. if confidence, it inverts. 

Possible improvement: change pixel_to_mm_scale to "scale_error_factor"